cloudferrylib/os/identity/keystone.py

```diff
@@ -474,7 +477,78 @@ def _get_user_tenants_roles(self, tenant_list=None, user_list=None):
             tenant_list = []
         if user_list is None:
             user_list = []
-        user_tenants_roles = {}
+        if not self.config.identity.optimize_user_role_fetch:
+            user_tenants_roles = self._get_user_tenants_roles_by_api(tenant_list,
+                                                                     user_list)
+        else:
+            user_tenants_roles = self._get_user_tenants_roles_by_db(tenant_list,
+                                                                    user_list)
+        return user_tenants_roles
+
+    def _get_roles_sql_request(self):
+        res = []
+        try:
+            is_project_metadata = self.mysql_connector.execute(
```

Comment:
this whole logic must be isolated in hierarchy of classes with the same interface and covered with unit tests.